src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -415,10 +483,22 @@ private void TrackNullableStateForAssignment(BoundNode node, Symbol assignmentTa
                             valueIsNotNull;
                     }
 
-                    if (slot > 0 && targetType.TypeSymbol.IsAnonymousType && targetType.TypeSymbol.IsClassType() &&
-                        (value == null || targetType.TypeSymbol == value.Type))
+                    if (slot > 0)
                     {
-                        InheritNullableStateOfAnonymousTypeInstance(targetType.TypeSymbol, slot, GetValueSlotForAssignment(value), isByRefTarget);
+                        InheritDefaultState(slot);
+
+                        // PROTOTYPE(NullableReferenceTypes): We're copying fields of anonymous type assignments,
+                        // even if the RHS is an alias. Is that the expected behavior or should we ignore aliases in this case?
+                        if (value != null &&
+                            (value.Kind == BoundKind.ObjectCreationExpression || value.Kind == BoundKind.AnonymousObjectCreationExpression || targetType.TypeSymbol.IsAnonymousType) &&
```

Comment:
>(value.Kind == BoundKind.ObjectCreationExpression || value.Kind == BoundKind.AnonymousObjectCreationExpression || targetType.TypeSymbol.IsAnonymousType) && [](start = 28, length = 155)

It doesn't feel right to have this condition. If we assigning a value, we should inherit all tracking information that we have about it, regardless its type or shape of the value.

Reply:
For example we don't have this restriction in InheritNullableStateOfFieldOrProperty.

---
In reply to: [152697434](https://github.com/dotnet/roslyn/pull/23063#discussion_r152697434) [](ancestors = 152697434)

Reply:
We probably should copy all track state regardless of value, but we'll need to handle cycles (for instance, the assignment to `C.F` below). For now, I've added skipped tests and left the `// PROTOTYPE` comments as is.
```
class C
{
    C? F;
    C()
    {
        F = this;
    }
}
```

Reply:
> We probably should copy all track state regardless of value, but we'll need to handle cycles (for instance, the assignment to C.F below). For now, I've added skipped tests and left the // PROTOTYPE comments as is.

Could you please clarify the comment with some details what the problem/challenge is and/or the "link" to the test?

Reply:
Updated comment.